File: sort/insert/insert_sort.py
#!/usr/bin/env python3

import random


def insert_sort(array):
    for j in range(1, len(array)):
        i = j
        key = array[i]
        # array[0:j-1] is sorted
        # insert array[i] into array[0:i-1]
        while i > 0 and key < array[i-1]:
            array[i] = array[i-1]
            i = i - 1

        array[i] = key


if __name__ == "__main__":
    array = [i for i in range(101)]
    random.shuffle(array)
    print(array)

    insert_sort(array)
    print(array)

File: sort/insert/insert_sort.js
const insertSort = array => {
    /*
     * sorted: array[0..i)
     * wait sort: array[i..length)
     * insert array[i] into array[0..i)
     */
    for (let i = 1; i < array.length; i++) {
        let j = i
        let tmp = array[i]
        while (j > 0 && array[j - 1] > tmp) {
            array[j] = array[j - 1]
            j--
        }
        array[j] = tmp
    }

    return array
}

module.exports = insertSort

if (require.main === module) {
    const readline = require('readline');

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    rl.question('请输入一串数字，用空格分隔：', (inputString) => {
        const numbersArray = inputString.split(' ').map(Number);
        console.log(numbersArray);

        sortedArray = insertSort(numbersArray)
        console.log(sortedArray);

        rl.close();
    });
}
